src/liquidity.rs

```diff
@@ -0,0 +1,418 @@
+				}
+			}
+
+			log_debug!(
+				self.logger,
+				"Choosing cheapest liquidity offer, will pay {}ppm msat in proportional LSP fees",
+				min_prop_fee_ppm_msat
+			);
+			(None, Some(min_prop_fee_ppm_msat), min_params)
+		};
+
+		let buy_response = self.send_buy_request(amount_msat, min_opening_params.clone()).await?;
+
+		// LSPS2 requires min_final_cltv_expiry_delta to be at least 2 more than usual.
+		let min_final_cltv_expiry_delta = MIN_FINAL_CLTV_EXPIRY_DELTA + 2;
+		let (payment_hash, payment_secret) = self
+			.channel_manager
+			.create_inbound_payment(None, expiry_secs, Some(min_final_cltv_expiry_delta))
+			.map_err(|e| {
+				log_error!(self.logger, "Failed to register inbound payment: {:?}", e);
+				Error::InvoiceCreationFailed
+			})?;
+
+		let route_hint = RouteHint(vec![RouteHintHop {
+			src_node_id: lsps2_service.node_id,
+			short_channel_id: buy_response.intercept_scid,
+			fees: RoutingFees { base_msat: 0, proportional_millionths: 0 },
+			cltv_expiry_delta: buy_response.cltv_expiry_delta as u16,
+			htlc_minimum_msat: None,
+			htlc_maximum_msat: None,
+		}]);
+
+		let payment_hash = sha256::Hash::from_slice(&payment_hash.0).map_err(|e| {
+			log_error!(self.logger, "Invalid payment hash: {:?}", e);
+			Error::InvoiceCreationFailed
+		})?;
+
+		let currency = self.config.network.into();
+		let mut invoice_builder = InvoiceBuilder::new(currency)
+			.description(description.to_string())
+			.payment_hash(payment_hash)
+			.payment_secret(payment_secret)
+			.current_timestamp()
+			.min_final_cltv_expiry_delta(min_final_cltv_expiry_delta.into())
+			.expiry_time(Duration::from_secs(expiry_secs.into()))
+			.private_route(route_hint);
+
+		if let Some(amount_msat) = amount_msat {
+			invoice_builder = invoice_builder.amount_milli_satoshis(amount_msat).basic_mpp();
+		}
+
+		let invoice = invoice_builder
+			.build_signed(|hash| {
+				Secp256k1::new()
+					.sign_ecdsa_recoverable(hash, &self.keys_manager.get_node_secret_key())
+			})
+			.map_err(|e| {
+				log_error!(self.logger, "Failed to build and sign invoice: {}", e);
+				Error::InvoiceCreationFailed
+			})?;
+
+		log_info!(self.logger, "JIT-channel invoice created: {}", invoice);
+		Ok((invoice, min_total_fee_msat, min_prop_fee_ppm_msat))
+	}
+
+	async fn request_opening_fee_params(&self) -> Result<LSPS2FeeResponse, Error> {
+		let lsps2_service = self.lsps2_service.as_ref().ok_or(Error::LiquiditySourceUnavailable)?;
+
+		let client_handler = self.liquidity_manager.lsps2_client_handler().ok_or_else(|| {
+			log_error!(self.logger, "Liquidity client was not configured.",);
+			Error::LiquiditySourceUnavailable
+		})?;
+
+		client_handler.request_opening_params(lsps2_service.node_id, lsps2_service.token.clone());
+
+		let mut fee_request_receiver = lsps2_service.pending_fee_request_receiver.lock().unwrap();
+		tokio::time::timeout(
+			Duration::from_secs(LIQUIDITY_REQUEST_TIMEOUT_SECS),
+			fee_request_receiver.recv(),
+		)
+		.await
```

Comment:
Could there be a race where `request_opening_fee_params` is called twice concurrently but the responses get mixed up? (e.g., Because of network latency the LSP receives and responds to the second request first.)